File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.java
import java.util.*;

class Solution {
    public int[] solution(int[] numbers) {
        int[] answer = new int[numbers.length];
        Arrays.fill(answer, -1);
        Stack<Integer> s = new Stack<>();
        
        for(int i = 0; i < numbers.length; i++) {
            while (!s.isEmpty() && numbers[s.peek()] < numbers[i]) {
                answer[s.pop()] = numbers[i];
            }
            s.push(i);
        }
        return answer;
    }
}

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.py
def solution(numbers):
    answer = [-1] * len(numbers)
    stack = []
    
    for i in range(len(numbers)):
        temp = numbers[i]
        
        while stack and numbers[stack[-1]] < temp:
            answer[stack.pop()] = temp
        
        stack.append(i)
    
    # for i in range(len(numbers)-1):
    #     for j in numbers[i+1:len(numbers)]:
    #         if numbers[i] < j:
    #             answer[i] = j
    #             break
        
    return answer